src/advanced/merkelized-map.md

```diff
@@ -0,0 +1,233 @@
+need to be transformed to the corresponding format.
+`ProofMapIndex` provides two ways to accomplish this:
+
+- **Hashed** transform is used by default and works with any serializable key type.
+  In this case, the transform is to hash the key serialization
+  with SHA-256 and use the output as an MPT path. Because of cryptographic
+  properties of SHA-256, the transform is resistant to collisions and pre-image
+  attacks.
+- **Raw** transform means that the original key serialization is used. This transform
+  is advised only for key types that always serialize to 32 bytes and have
+  uniform distribution over the key space. (The latter is to limit the MPT depth.)
+  Types that can be used with the raw transform include 32-byte outputs
+  of a cryptographic hash function and Ed25519 public keys.
+
+Because the key transform may be irreversible, `ProofMapIndex` stores both original
+map keys and paths in the MPT split with a prefix tag. Iteration and retrieval
+are always performed on the original keys.
+
+### Redefining Tree Root
+
+Because of implementation reasons, nodes in an MPT used in MerkleDB can have
+0 or 2 children, never 1. This requirement naturally follows for leaves
+and intermediate nodes in a generic binary trie; it can only be violated
+for the root node (e.g., if all paths in the trie start with `0` bit).
+
+To deal with this, MerkleDB redefines the tree root as the only child
+node of a “true” root node if the latter has a single child.
+Thus, in the following description the tree root always
+has 0 or 2 children, and there is a special case if the MPT
+has a single leaf node.
+
+## Computing Map Hash
+
+[As with lists](merkelized-list.md#computing-list-hash), computing hash
+of a map consists of two steps:
+
+1. Computing the *root hash* of the MPT
+2. Computing the *map hash* using the root hash.
+
+The second part is easy; given the `root_hash` of the MPT, the hash
+of the corresponding map is
+
+```text
+map_hash = hash( HashTag::MapNode || root_hash ),
+```
+
+where `HashTag::MapNode = 3` is the domain separation tag for map objects,
+`hash()` is the SHA-256 function, `||` is byte concatenation.
+(Here and elsewhere, tags are serialized as a single byte.)
+
+How `root_hash` computed, depends on the number of entries in the map.
+
+### Empty Map
+
+Root hash of an empty map is defined as 32 zero bytes.
+
+### Map With a Single Element
+
+Root hash for a map with a single entry is computed as follows:
+
+```text
+root_hash = hash(
+    HashTag::MapBranchNode ||
+    0x01 || key_path || 0x00 ||
+    value_hash
+);
+value_hash = hash(HashTag::Blob || value),
+```
+
+where
+
+- `HashTag::MapBranchNode = 4` is the domain separation tag for MPT nodes.
+- `key_path` is the 256-bit MPT path obtained by transforming the entry key.
+- `value_hash` is the domain-separated hash of the entry value
+  with tag `HashTag::Blob = 0`.
+
+`0x01` and `0x00` are single bytes present for historic reasons.
```

Comment:
Don't quite like this.

Reply:
Totally agree. I even created a task for that ([ECR-3485 Merkledb: Unify keys hashing in MapProofIndex](https://jira.bf.local/browse/ECR-3485)) a while ago, but had no time to implement it.